### PR TITLE
Support transaction signing by heterogenous lists of keypairs

### PR DESCRIPTION
**Pull request by garious**
#### Problem

Transactions can only be signed by a slice of keypairs that implement the same instance of `KeypairUtil`.  Previous attempts to generalize this touched a boatload of code and introduced dynamic allocations and dynamic dispatch.

#### Summary of Changes

* Create a new trait `Keypairs`. Implement it for the previously-used types.
* Add instance for heterogeneous signers, via `Box<dyn KeypairUtil>`

### CODE REVIEW

**Comment by garious** [comment]
@CriesofCarrots, check it out

**Comment by CriesofCarrots** [comment]
Yep, I can work with this, although having to re-implement for each newly desired type of collection makes me sad.
We'll need a `impl<T: KeypairUtil, U: KeypairUtil, V: KeypairUtil, W: KeypairUtil> Keypairs for (T, U, V, W)`.
We do also need a fallible sign_message (I was calling it `try_sign_message`) for some implementers of KeypairUtil, so I guess that will need to be added to KeypairUtil and Keypairs both

**Comment by garious** [comment]
Instead of creating new large tuple instances how about either creating a generic instance for `[Box<dyn KeypairUtil>]` or implement the trait from application-specific structs?

**Comment by CriesofCarrots** [comment]
> Instead of creating new large tuple instances how about either creating a generic instance for `[Box<dyn KeypairUtil>]` or implement the trait from application-specific structs?

I'm confused. Wasn't your whole purpose to avoid `Box<dyn KeypairUtil>`?

**Comment by CriesofCarrots** [comment]
... if not, I would run with it, as it makes parsing and deduplicating signer inputs of various types more straightforward...

File: sdk/src/keypairs.rs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    fn pubkeys(&self) -> Vec<Pubkey> {
+        vec![self.0.pubkey(), self.1.pubkey()]
+    }
+
+    fn sign_message(&self, message: &[u8]) -> Vec<Signature> {
+        vec![self.0.sign_message(message), self.1.sign_message(message)]
+    }
+}
+
+impl<T: KeypairUtil, U: KeypairUtil, V: KeypairUtil> Keypairs for (T, U, V) {
```
Comment (jstarry) [inline sdk/src/keypairs.rs]: macros work well for this

**Comment by garious** [comment]
No, my whole point was not to avoid `Box<dyn KeypairUtil>`. It was to avoid changing things at the lowest levels that would force `Box<dyn KeypairUtil>` onto all users.  If it's the right tool for the job, then go ahead and use it.

That said, I'm having plenty of trouble with this PR. The Rust type system isn't letting me make instances as generic as I'd like.

File: sdk/src/lib.rs [hunk 2]
```diff
@@ -12,6 +12,7 @@ pub mod fee_calculator;
 pub mod hash;
 pub mod inflation;
 pub mod instruction;
+pub mod keypairs;
```
Comment (CriesofCarrots) [inline sdk/src/lib.rs]: You just need to move this below as:
```
#[cfg(not(feature = "program"))]
pub mod keypairs;
```
to pass CI checks

Comment (garious) [reply sdk/src/lib.rs]: Thanks!

File: sdk/src/keypairs.rs (hunk 1 above)
Comment (garious) [reply sdk/src/keypairs.rs]: Wrote a terrible one to dedup the code, thanks.  Unfortunately, no time to make a macro to invoke that macro ~10 times.

**Comment by garious** [comment]
@jstarry, @CriesofCarrots, @t-nelson, this is the best I've been able to manage. It aims to solve only the problem that `<T: KeypairUtil> &[T]` works on just one type of `KeypairUtil` and we now have concrete cases where one keypair might be a `Keypair` while another might be a hardware wallet.

Regarding naming, I don't like `Keypairs` at all.  I saw some discussion in the other PRs, but don't know if there was consensus.  Happy to merge this one with `Keypairs` to get @CriesofCarrots unblocked, or do a rename to something like `Signers` now.  Thoughts?

**Comment by CriesofCarrots** [comment]
> @jstarry, @CriesofCarrots, @t-nelson, this is the best I've been able to manage. It aims to solve only the problem that `<T: KeypairUtil> &[T]` works on just one type of `KeypairUtil` and we now have concrete cases where one keypair might be a `Keypair` while another might be a hardware wallet.
> 
> Regarding naming, I don't like `Keypairs` at all. I saw some discussion in the other PRs, but don't know if there was consensus. Happy to merge this one with `Keypairs` to get @CriesofCarrots unblocked, or do a rename to something like `Signers` now. Thoughts?

My vote is for Signers

**Comment by t-nelson** [comment]
This looks good to me!  Especially with the tuple implementations out. It occurred to me that, after @CriesofCarrots's signer de-dup lands, we likely won't know until run time how many signers there are for any particular use.

I prefer `Signers` to `Keypairs` as well.  Indifferent as to now or later, though

**Comment by garious** [comment]
okay, switching to Signers

**Comment by garious** [comment]
automerge label engaged

**Comment by solana-grimes** [comment]
:broken_heart: Unable to automerge due to CI failure

**Pull request merged**